### PR TITLE
Implement OP_SHA1 opcode with tests

### PR DESCRIPTION
**Pull request by supreme2580**
Implement OP_SHA1 opcode with tests

Closes: https://github.com/zig-bitcoin/btczee/issues/67

### CODE REVIEW

File: src/script/engine.zig [hunk 1]
```diff
@@ -476,29 +478,39 @@ pub const Engine = struct {
...
-};
-
-test "Script execution - OP_1 OP_1 OP_EQUAL" {
-    const allocator = std.testing.allocator;
 
-    // Simple script: OP_1 OP_1 OP_EQUAL
-    const script_bytes = [_]u8{ 0x51, 0x51, 0x87 };
-    const script = Script.init(&script_bytes);
+    fn opSha1(self: *Engine) !void {
+        const data = try self.stack.pop();
```
Comment (tdelabro) [inline src/script/engine.zig:483]: defer self.allocator.free(data)

File: src/script/engine.zig [hunk 2]
```diff
@@ -476,29 +478,39 @@ pub const Engine = struct {
...
 
-    try engine.execute();
+        var actual_output: [sha1.digest_length]u8 = undefined;
 
-    // Check if the execution result is true (non-empty stack with top element true)
-    {
-        const result = try engine.stack.pop();
-        defer allocator.free(result);
-        try std.testing.expect(result.len > 0 and result[0] != 0);
+        sha1.hash(case.input, &actual_output, .{});
```
Comment (tdelabro) [inline src/script/engine.zig:511]: you are writing a test for sha1, which is part of the zig language std lib.
You should be writing a test for `opSha1`, the function you wrote

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: src/script/engine.zig (hunk 2 above)
Comment (supreme2580) [reply src/script/engine.zig:511]: Oh... will fix that too😁

**Comment by supreme2580** [comment]
@tdelabro I have implemented the requested changes

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: src/script/engine.zig [hunk 3]
```diff
@@ -476,29 +478,54 @@ pub const Engine = struct {
...
 
-    // Simple script: OP_1 OP_1 OP_EQUAL
-    const script_bytes = [_]u8{ 0x51, 0x51, 0x87 };
-    const script = Script.init(&script_bytes);
+    for (test_cases) |case| {
+        const allocator = std.testing.allocator;
 
-    var engine = Engine.init(allocator, script, .{});
-    defer engine.deinit();
+        const script_bytes = [_]u8{0xa7};
```
[the remark this answers is not in the document]
Comment (supreme2580) [reply src/script/engine.zig:506]: Will do boss

[the remark this answers is not in the document]
Comment (supreme2580) [reply src/script/engine.zig]: @tdelabro done boss

File: src/script/engine.zig [hunk 4]
```diff
@@ -476,29 +478,54 @@ pub const Engine = struct {
...
+        var engine = Engine.init(allocator, script, .{});
+        defer engine.deinit();
 
-    // Check if the execution result is true (non-empty stack with top element true)
-    {
+        // Push the input onto the stack
+        try engine.stack.pushByteArray(case.input);
+
+        // Call opSha1
+        try engine.opSha1();
```
[the remark this answers is not in the document]
Comment (supreme2580) [reply src/script/engine.zig:516]: Changing it right now

[the remark this answers is not in the document]
Comment (supreme2580) [reply src/script/engine.zig]: @tdelabro done boss

**Review by tdelabro — APPROVED** [review]

**Pull request merged**